### PR TITLE
TEP-0121: Adds a new field `Retries` for TaskRun

### PR DESCRIPTION
**Pull request by XinruZhang**
# Changes

This PR adds a new field `Retries` for `v1` and `v1beta1` TaskRun, which enables TaskRun reconciler implements the `Retries` independently from PipelineRun reconciler.

This PR also adds a new member function `IsRetriable` for TaskRun to indicate whether `taskRun.Retries` has been exhausted or not.

issue: #5756 

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Has a kind label. You can add one by adding a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [x] Release notes block below has been updated with any user facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [x] ~Release notes contains the string "action required" if the change requires additional action from users switching to the new release~

# Release Notes

```release-note
API change: new `Retries` field in `v1beta1.TaskRun` and `v1.TaskRun`.
```

### CODE REVIEW

**Comment by XinruZhang** [comment]
This PR is the first peice of https://github.com/tektoncd/pipeline/pull/5842 adding API changes, will open another PR for adding the reconciler logic. We targeting at getting both in before cutting the release.

File: pkg/apis/pipeline/v1/taskrun_types.go [hunk 1]
```diff
@@ -141,9 +145,11 @@ const (
 	TaskRunReasonSuccessful TaskRunReason = "Succeeded"
 	// TaskRunReasonFailed is the reason set when the TaskRun completed with a failure
 	TaskRunReasonFailed TaskRunReason = "Failed"
-	// TaskRunReasonCancelled is the reason set when the Taskrun is cancelled by the user
+	// TaskRunReasonToBeRetried is the reason set when the last TaskRun execution failed, and will be retried
+	TaskRunReasonToBeRetried TaskRunReason = "ToBeRetried"
+	// TaskRunReasonCancelled is the reason set when the TaskRun is cancelled by the user
 	TaskRunReasonCancelled TaskRunReason = "TaskRunCancelled"
-	// TaskRunReasonTimedOut is the reason set when the Taskrun has timed out
+	// TaskRunReasonTimedOut is the reason set when the TaskRun has timed out
```
Comment (chitrangpatel) [inline pkg/apis/pipeline/v1/taskrun_types.go:152]: May be make it clear if its all retries have timed out or a retry attempt has timed out?

File: pkg/apis/pipeline/v1beta1/taskrun_types.go [hunk 2]
```diff
@@ -166,14 +169,16 @@ const (
 	TaskRunReasonSuccessful TaskRunReason = "Succeeded"
 	// TaskRunReasonFailed is the reason set when the TaskRun completed with a failure
 	TaskRunReasonFailed TaskRunReason = "Failed"
-	// TaskRunReasonCancelled is the reason set when the Taskrun is cancelled by the user
+	// TaskRunReasonToBeRetried is the reason set when the last TaskRun execution failed, and will be retried
+	TaskRunReasonToBeRetried TaskRunReason = "ToBeRetried"
+	// TaskRunReasonCancelled is the reason set when the TaskRun is cancelled by the user
 	TaskRunReasonCancelled TaskRunReason = "TaskRunCancelled"
-	// TaskRunReasonTimedOut is the reason set when the Taskrun has timed out
+	// TaskRunReasonTimedOut is the reason set when the TaskRun has timed out
```
Comment (chitrangpatel) [inline pkg/apis/pipeline/v1beta1/taskrun_types.go:176]: same comment as above.

**Review by chitrangpatel — APPROVED** [review]

File: pkg/apis/pipeline/v1/taskrun_types.go (hunk 1 above)
Comment (XinruZhang) [reply pkg/apis/pipeline/v1/taskrun_types.go]: SG, updated.

File: pkg/apis/pipeline/v1beta1/taskrun_types.go (hunk 2 above)
Comment (XinruZhang) [reply pkg/apis/pipeline/v1beta1/taskrun_types.go]: done.

File: docs/taskruns.md [hunk 3]
```diff
@@ -698,11 +699,25 @@ object(s), if present. Any `Request` or `Limit` specified by the user (on `Task`
...
+- `status.StartTime` and `status.PodName` are unset to trigger another retry attempt.
+
 ### Configuring the failure timeout
 
-You can use the `timeout` field to set the `TaskRun's` desired timeout value. If you do not specify this
-value for the `TaskRun`, the global default timeout value applies. If you set the timeout to 0, the `TaskRun` will
-have no timeout and will run until it completes successfully or fails from an error.
+You can use the `timeout` field to set the `TaskRun's` desired timeout value for **each retry attempt**. If you do
+not specify this value, the global default timeout value applies (the same, to `each retry attempt`). If you set the timeout to 0,
+the `TaskRun` will have no timeout and will run until it completes successfully or fails from an error.
```
Comment (jerop) [inline docs/taskruns.md:720]: let's remove these docs for now, and add them in the final pull request so that we don't confuse users of nightly releases who may think that it's fully supported -- see similar comments from @afrittoli on a recent PR: https://github.com/tektoncd/pipeline/pull/5828#discussion_r1042336732

File: pkg/apis/pipeline/v1beta1/taskrun_types.go [hunk 4]
```diff
@@ -256,13 +261,13 @@ type TaskRunStatusFields struct {
...
 
-	// Results from Resources built during the taskRun. currently includes
+	// Results from Resources built during the TaskRun. currently includes
 	// the digest of build container images
 	// +optional
 	// +listType=atomic
 	ResourcesResult []PipelineResourceResult `json:"resourcesResult,omitempty"`
 
-	// TaskRunResults are the list of results written out by the task's containers
+	// TaskRunResults are the list of results written out by the Task's containers
```
Comment (jerop) [inline pkg/apis/pipeline/v1beta1/taskrun_types.go:270]: there's a bunch of these cleanup for unrelated code - can that please go in a separate commit and PR?

Comment (XinruZhang) [reply pkg/apis/pipeline/v1beta1/taskrun_types.go]: Thanks Jerop, updated.

File: docs/taskruns.md (hunk 3 above)
Comment (XinruZhang) [reply docs/taskruns.md]: SG, removed!

**Review by jerop — APPROVED** [review]
Thank you @XinruZhang!
